include/envoy/config/extension_config_provider.h

```diff
@@ -0,0 +1,51 @@
+#pragma once
+
+#include "envoy/common/pure.h"
+
+#include "common/protobuf/protobuf.h"
+
+#include "absl/types/optional.h"
+
+namespace Envoy {
+namespace Config {
+
+/**
+ * A provider for extension configurations obtained either statically or via
+ * the extension configuration discovery service. Dynamically updated extension
+ * configurations may share subscriptions across extension config providers.
+ */
+template <class Factory, class FactoryCallback> class ExtensionConfigProvider {
+public:
+  virtual ~ExtensionConfigProvider() = default;
+
+  /**
+   * Get the extension configuration resource name.
+   **/
+  virtual const std::string& name() PURE;
+
+  /**
+   * @return FactoryCallback an extension factory callback to be applied on subsequent streams.
```

Comment:
nit: as this is now generic I would make the verbiage around "streams" more generic. Same below.

Reply:
Updated.